src/lib.rs

```diff
@@ -1,14 +1,267 @@
+            .read_u24::<LittleEndian>()
+            .context("failed to read u24")?;
+        let as_u8: u8 = value.try_into().context("decoded value too big for u8")?;
+        Self::try_from(as_u8)
+    }
+
+    fn encode(&self, bytes: &mut Vec<u8>) -> Result<(), anyhow::Error> {
+        bytes
+            .write_u24::<LittleEndian>(*self as u32)
+            .context("failed to write u24")
+    }
+}
+
+/// FieldP128 is the field with modulus 2^128 - 2^108 + 1, described in [Section 7.2 of
+/// draft-google-cfrg-libzk-00][1]. The field does not get a name in the draft, but P128 comes from
+/// the longfellow implementation ([3]).
+///
+/// The generator was computed in SageMath as `GF(2^128-2^108+1).primitive_element()` (thanks to the
+/// hint in [`PrimeField::MULTIPLICATIVE_GENERATOR`]).
+///
+/// The endianness is per [Section 7.2.1 of draft-google-cfrg-libzk-00][2].
+///
+/// [1]: https://www.ietf.org/id/draft-google-cfrg-libzk-00.html#section-7.2
+/// [2]: https://www.ietf.org/id/draft-google-cfrg-libzk-00.html#section-7.2.1
+/// [3]: https://github.com/google/longfellow-zk/blob/main/lib/algebra/fp_p128.h
+#[derive(ff::PrimeField)]
+#[PrimeFieldModulus = "340282042402384805036647824275747635201"]
+#[PrimeFieldGenerator = "59"]
+#[PrimeFieldReprEndianness = "little"]
+// ff requires that the repr be an array of u64 and despite the fact that 128 bits should be big
+// enough, also requires 3 u64s.
+struct FieldP128([u64; 3]);
+
+/// A serialized size, which is in the range [1, 2^24 -1] per [draft-google-cfrg-libzk-00 section
+/// 7][1]. Serialized in little endian order, occupying 3 bytes.
+#[derive(Clone, Copy, Debug, Eq, PartialEq, PartialOrd, Ord, Default)]
+pub struct Size(u32);
+
+impl From<u32> for Size {
+    fn from(value: u32) -> Self {
+        Self(value)
+    }
+}
+
+impl From<Size> for usize {
+    fn from(value: Size) -> Self {
+        // XXX shouldn't assume that usize is big enough for u32
+        value.0 as Self
+    }
+}
+
+impl Codec for Size {
+    fn decode(bytes: &mut Cursor<&[u8]>) -> Result<Self, anyhow::Error> {
+        Ok(Self(
+            bytes
+                .read_u24::<LittleEndian>()
+                .context("failed to read u24")?,
+        ))
+    }
+
+    fn encode(&self, bytes: &mut Vec<u8>) -> Result<(), anyhow::Error> {
+        if self.0 > 1 << 24 {
+            return Err(anyhow!(
+                "size {} too big to be serialized in 3 bytes",
+                self.0
+            ));
+        }
+        bytes
+            .write_u24::<LittleEndian>(self.0)
+            .context("failed to write u24")
+    }
+}
+
+impl Size {
+    /// Encode this value as a delta from the previous value in some sequence.
+    pub fn encode_delta(&self, previous: Size, bytes: &mut Vec<u8>) -> Result<(), anyhow::Error> {
+        let delta = if self.0 >= previous.0 {
+            // Delta is positive: shift the delta up by one, leaving sign bit clear
+            (self.0 - previous.0) << 1
```

Comment:
These shifts should probably have checks that they don't wrap around, or the circuit would get corrupted.